src/crypto/tls/common.go

```diff
@@ -747,6 +729,11 @@ type Config struct {
 	// payload.
 	ServerECHProvider ECHProvider
 
+	// EventHandler, if set, is called by the client and server at various
+	// points during the handshake to handle specific events. This callback can
+	// be used to record interesting metrics.
+	EventHandler func(event string)
```

Comment:
I'd imagine we want something more general than a `string` for the event's type. This might be better:
```
type Event interface {
    String() string
}
```

Reply:
Yeah, or perhaps even: 

```go
type Event interface {
   Type() int
   Name() string
}

type CounterEvent struct {
   Count int
}

... to process an event ... 
if event.Type() == CounterEvent {
   counterEvent, ok := event.(CounterEvent)
   if !ok { panic(...) }
   // use counterEvent as needed
}
```

Sort of like [BoringSSL's info callback mechanism](https://github.com/google/boringssl/blob/master/include/openssl/ssl.h#L3976)

Reply:
You don't need .Type: it's possible to introspect based on types in go.

Reply:
Whatever's most idiomatic :) If you want to cast-then-check, that works too.

Reply:
I went with
```
type Event interface {
   Name() string
}
```
I don't see a reason to do anything fancier. (We can always change this later.)